hibernate-entitymanager/src/test/java/org/hibernate/jpa/test/EntityManagerFactoryUnwrapTest.java

```diff
@@ -0,0 +1,135 @@
+
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.PersistenceException;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import org.hibernate.SessionFactory;
+import org.hibernate.ejb.HibernateEntityManagerFactory;
+import org.hibernate.engine.spi.SessionFactoryImplementor;
+import org.hibernate.internal.SessionFactoryImpl;
+import org.hibernate.jpa.internal.EntityManagerFactoryImpl;
+import org.hibernate.jpa.internal.EntityManagerImpl;
+import org.hibernate.testing.TestForIssue;
+
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.fail;
+
+/**
+ * Test various unwrap scenarios for {@code EntityManagerFactory}.
+ */
+@TestForIssue(jiraKey = "HHH-9665")
+public class EntityManagerFactoryUnwrapTest extends BaseEntityManagerFunctionalTestCase {
+	private EntityManagerFactory entityManagerFactory;
+
+	@Override
+	public Class[] getAnnotatedClasses() {
+		return new Class[] { };
+	}
+
+	@Before
+	public void setUp() {
+		entityManagerFactory = entityManagerFactory();
+	}
+
+	@Test
+	public void testEntityManagerCanBeUnwrappedToSessionFactory() {
+		SessionFactory sessionFactory = entityManagerFactory.unwrap( SessionFactory.class );
+		assertNotNull( "Unwrapping to API class SessionFactory should be ok", sessionFactory );
+	}
+
+	@Test
+	public void testEntityManagerCanBeUnwrappedToSessionFactoryImplementor() {
+		SessionFactoryImplementor sessionFactoryImplementor = entityManagerFactory.unwrap( SessionFactoryImplementor.class );
+		assertNotNull( "Unwrapping to SPI class SessionFactoryImplementor should be ok", sessionFactoryImplementor );
+	}
+
+	@Test
+	public void testEntityManagerCanBeUnwrappedToDeprecatedHibernateEntityManagerFactory() {
+		HibernateEntityManagerFactory hibernateEntityManagerFactory = entityManagerFactory.unwrap(
+				HibernateEntityManagerFactory.class
+		);
+		assertNotNull(
+				"Unwrapping to SPI class HibernateEntityManagerFactory should be ok",
+				hibernateEntityManagerFactory
+		);
+	}
+
+	@Test
+	public void testEntityManagerCanBeUnwrappedToHibernateEntityManagerFactory() {
+		org.hibernate.jpa.HibernateEntityManagerFactory hibernateEntityManagerFactory = entityManagerFactory.unwrap( org.hibernate.jpa.HibernateEntityManagerFactory.class );
+		assertNotNull(
+				"Unwrapping to SPI class HibernateEntityManagerFactory should be ok",
+				hibernateEntityManagerFactory
+		);
+	}
+
+	@Test
+	public void testEntityManagerCanBeUnwrappedToObject() {
+		Object object = entityManagerFactory.unwrap( Object.class );
+		assertNotNull( "Unwrapping to public super type Object should work", object );
+	}
+
+	@Test
+	public void testEntityManagerCanBeUnwrappedToSessionFactoryImpl() {
+		SessionFactoryImpl sessionFactory = entityManagerFactory.unwrap( SessionFactoryImpl.class );
+		assertNotNull(
+				"Unwrapping to SessionFactoryImpl should be ok",
+				sessionFactory
+		);
+	}
+
+	@Test
+	public void testEntityManagerCanBeUnwrappedToEntityManagerFactoryImpl() {
```

Comment:
Ok, you just did not update the test.  In fact this test fails.  I'll correct it.